File: shortest_path/cxx/bfs.cxx
#include <deque>
#include <unordered_set>
#include <vector>

// UNTESTED

// Nodes are 0 .. n-1 by default
void bfs(std::vector<std::vector<int> > adjList)
{
	std::deque<int> front(1, 0);
	std::unordered_set<int> visited;
	while(!front.empty())
	{
		int frontLen = front.size();
		for(int i = 0; i < frontLen; i++)
		{
			int node = front.front();
			front.pop_front();
			for(const int& child: adjList[node])
				if(visited.find(child) == visited.end())
				{
					front.push_back(child);
					visited.insert(child);
				}
		}
	}
}

File: shortest_path/py/bfs.py
from collections import deque

# Tested
def bfs(adjList, source):
    front = deque([source])
    dist = {source: 0}
    prev = {source: None}
    while front:
        node = front.popleft()
        for child in adjList[node]:
            if child not in dist:
                front.append(child)
                dist[child] = dist[node] + 1
                prev[child] = node
    return dist, prev

def bfs_template_1(adjList, source):
    front = deque([source])
    visited = {source}
    while front:
        node = front.popleft()
        # Do something here
        for child in adjList[node]:
            if child not in visited:
                front.append(child)
                visited.add(child)
    return None

# Verified
def bfs_template_2(adjList, source):
    front = [source]
    visited = {source}
    while front:
        children = []
        for node in front:
            # Do something here
            for child in adjList[node]:
                if child not in visited:
                    children.append(child)
                    visited.add(child)
        front = children
    return None
